Remove backward op for slow 3d convolution

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69978

There is no CUDA implementation for slow 3d, so this PR just calls the CPU kernel directly. This avoids another dispatch entirely, speeding up the logic somewhat.

Differential Revision: [D33131003](https://our.internmc.facebook.com/intern/diff/D33131003)

jbschlosser committed:
Remove backward op for slow 3d convolution

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131003).

jbschlosser committed:
Update on "Remove backward op for slow 3d convolution"


Differential Revision: [D33131003](https://our.internmc.facebook.com/intern/diff/D33131003)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131003).

zou3519 commented on aten/src/ATen/native/cpu/ConvolutionMM3d.cpp:
```diff
@@ -0,0 +1,335 @@
+
+} // namespace
+
+REGISTER_DISPATCH(slow_conv3d_backward_stub, &slow_conv3d_backward_cpu);
```
I thought about it some more, and couldn't we do something like the following?
- keep slow_conv3d_backward_cpu in aten/native (and not in aten/native/cpu)
- Add a slow_conv3d_backward_stub that only has a CUDA implementation attached to it (and maybe a dummy CPU implementation if that's necessary)
- When we need to call "slow_conv3d_backward" from aten/native, first we check if the inputs are on CPU. If they are on CPU, then we invoke slow_conv3d_backward_cpu.
- If the inputs are not on CPU, then we invoke "slow_conv3d_backward_stub", which will call out to the CUDA version

This avoids any problems we have with compiling a file multiple times in native/cpu. Also, the consequences of unnecessarily putting something in native/cpu is more than just increased build times -- it is also larger binary sizes

jbschlosser replied:
Cool good call, and it's even easier for the specific case of slow 3d because there's no CUDA kernel for it :)

jbschlosser replied:
@zou3519 I've got a slightly different idea for avoiding recompilations: what if we introduce a new macro `REGISTER_ALL_CPU_DISPATCH` that registers the same kernel across all CPU arch types? We'd call this from `native/Convolution*.cpp` and wouldn't need to move any logic underneath the `native/cpu` dir. That would simplify these PRs quite a bit and would also avoid the recompilation. Wdyt about this approach?

zou3519 replied:
Just to check... for this specific case we don't need a dispatch macro, right? Because there is no CUDA implementation? If so you might want to change the PR body

jbschlosser replied:
Yep that's right! will change

jbschlosser committed:
Update on "Remove backward op for slow 3d convolution"


**Note:** `REGISTER_DISPATCH` for the CPU kernel is only accessible from the `native/cpu` directory. So this PR splits `aten/src/ATen/native/ConvolutionMM3d.cpp` into:
* (new file) `aten/src/ATen/native/ConvolutionMM3d.h` (contains functions shared between forward and backward)
* `aten/src/ATen/native/ConvolutionMM3d.cpp` (contains forward logic)
* (new file) `aten/src/ATen/native/cpu/ConvolutionMM3d.cpp` (contains backward functions + `REGISTER_DISPATCH` call)

Once the forward op is removed as well, the first two can go away.

Also note that there is no CUDA implementation for slow 3d.

Differential Revision: [D33131003](https://our.internmc.facebook.com/intern/diff/D33131003)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131003).

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -2046,10 +2046,7 @@
-
-- name: slow_conv3d_backward.output_mask(Tensor grad_output, Tensor self, Tensor weight, int[3] kernel_size, int[3] stride, int[3] padding, bool[3] output_mask) -> (Tensor grad_input, Tensor grad_weight, Tensor grad_bias)
-  grad_output, self, weight: _convolution_double_backward(grads[0], grads[1], grads[2], grad_output, weight, self, stride, padding, {{1, 1, 1}}, false, {{0, 0, 0}}, 1, false, false, false, false, grad_input_mask)
+  self, weight, bias: "grad.defined() ? convolution_backward(grad, self, weight, bias->sizes(), stride, padding, std::vector<int64_t>(padding.size(), 1), false, std::vector<int64_t>(padding.size(), 0), 1, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
```
std::vector<int64_t>(padding.size(), 0) -> /*output_padding*/{0, 0, 0} to avoid that dynamic allocation and make it more readable

zou3519 reviewed (approved):
You've sped up the operator by removing the dispatch and not adding a new one :D

jbschlosser committed:
Update on "Remove backward op for slow 3d convolution"


There is no CUDA implementation for slow 3d, so this PR just calls the CPU kernel directly. This avoids another dispatch entirely, speeding up the logic somewhat.

Differential Revision: [D33131003](https://our.internmc.facebook.com/intern/diff/D33131003)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33131003).
